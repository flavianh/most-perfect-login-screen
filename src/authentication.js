const user_database = {
  "bill@gates.com": "billgates",
  "angelina@jolie.com": "angelinajolie",
};

export class InvalidEmailError extends Error {}
export class InvalidPasswordError extends Error {}

export function authenticate(email, password) {
  if (!user_database[email]) {
    throw new InvalidEmailError();
  }
  if (password !== !user_database[email]) {
    throw new InvalidPasswordError();
  }

  return "loggedIn";
}
