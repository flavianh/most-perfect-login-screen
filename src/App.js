import * as Yup from "yup";

import { ErrorMessage, Field, Form, Formik } from "formik";
import { InvalidEmailError, InvalidPasswordError } from "./authentication";
import React, { useState } from "react";
import { Trans, Translation } from "react-i18next";

const LoginSchema = Yup.object().shape({
  email: Yup.string()
    .required("pleaseEnterEmail")
    .email("pleaseEnterValidEmail"),
  password: Yup.string().required("pleaseEnterPassword"),
});

function divTrans({ children, ...props }) {
  return (
    <div {...props}>
      <Trans>{children}</Trans>
    </div>
  );
}

function App(props) {
  const [logState, setLogState] = useState("loggedOut");
  const { authenticate } = props;

  return (
    <Translation>
      {(t) => (
        <Formik
          initialValues={{
            email: "", // Otherwise you get "cannot read property email of undefined"
            password: "",
          }}
          validationSchema={LoginSchema}
          onSubmit={({ email, password }, { setErrors }) => {
            try {
              authenticate(email, password);
              setLogState("loggedIn");
            } catch (error) {
              if (error instanceof InvalidEmailError) {
                setErrors({ email: "emailNotInOurSystem" });
              } else if (error instanceof InvalidPasswordError) {
                setErrors({ password: "passwordNotInOurSystem" });
              } else {
                setLogState("systemError");
              }
            }
          }}
        >
          {() => (
            <Form
              className="pure-form pure-form-stacked"
              aria-label="loginForm"
            >
              <div className="pure-control-group">
                <label htmlFor="email">{t("email")}</label>
                <Field id="email" name="email" type="email" required />
                <ErrorMessage
                  id="email-error"
                  role="alert"
                  component={divTrans}
                  className="pure-form-message-inline error-message"
                  name="email"
                />
              </div>
              <div className="pure-control-group">
                <label htmlFor="password">{t("password")}</label>
                <Field id="password" name="password" type="password" required />
                <ErrorMessage
                  component={divTrans}
                  id="password-error"
                  name="password"
                  role="alert"
                  className="pure-form-message-inline error-message"
                />
              </div>
              <div className="pure-controls">
                <button
                  type="submit"
                  className="pure-button pure-button-primary"
                >
                  {t("login")}
                </button>
              </div>
              <div id="log-state">{t(logState)}</div>
            </Form>
          )}
        </Formik>
      )}
    </Translation>
  );
}

export default App;
