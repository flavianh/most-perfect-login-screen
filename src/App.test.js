import { InvalidEmailError, InvalidPasswordError } from "./authentication";
import { fireEvent, render, wait } from "@testing-library/react";

import App from "./App";
import React from "react";

async function changeField(field, value) {
  await wait(() => {
    fireEvent.change(field, {
      target: {
        value,
      },
    });
  });
}

async function clickOnButton(button) {
  await wait(() => {
    fireEvent.click(button);
  });
}

test("Renders correctly", () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});

test("Logging state is loggedOut and form fields are empty by default", () => {
  const { getByText, getByLabelText } = render(<App />);
  const form = getByLabelText("loginForm");

  expect(getByLabelText("email").type).toBe("email");
  expect(getByLabelText("password").type).toBe("password");
  expect(getByText("loggedOut")).toBeInTheDocument();
  expect(form).toHaveFormValues({
    email: "",
    password: "",
  }); // This passes only if 'name' is set for form inputs
});

test("Logging state switches to loggedIn with valid credentials", async () => {
  const { getByLabelText, getByText } = render(
    <App
      authenticate={(email, password) => {
        if (email !== "testemail@email.com" || password !== "testpassword") {
          throw new Error("Invalid credentials");
        }
        return "loggedIn";
      }}
    />
  );
  const email = getByLabelText("email");
  const password = getByLabelText("password");
  const loginButton = getByText("login");

  await changeField(email, "testemail@email.com");
  await changeField(password, "testpassword");
  await clickOnButton(loginButton);

  expect(getByText("loggedIn")).toBeInTheDocument();
});

[
  { labelText: "email", errorMessage: "pleaseEnterEmail" },
  { labelText: "password", errorMessage: "pleaseEnterPassword" },
].forEach(({ labelText, errorMessage }) => {
  test(`Error ${errorMessage} should appear if ${labelText} was not touched when hitting Submit`, async () => {
    const { getByText, getByLabelText } = render(<App />);
    const loginButton = getByText("login");
    expect(getByLabelText(labelText)).toBeRequired();
    await clickOnButton(loginButton);
    const error = getByText(errorMessage);
    expect(error).toBeInTheDocument(); // Only works if "component" in ErrorMessage is set to something like div
    expect(error.id).toBe(`${labelText}-error`);
  });
});

test("pleaseEnterValidEmail should appear if a wrong email was set when hitting Submit", async () => {
  const { getByText, getByLabelText } = render(<App />);
  const labelText = "email";
  const email = getByLabelText(labelText);

  const loginButton = getByText("login");
  await changeField(email, "notanemail");
  await clickOnButton(loginButton);
  const error = getByText("pleaseEnterValidEmail");
  expect(error).toBeInTheDocument(); // Only works if "component" in ErrorMessage is set to something like div
  expect(error.id).toBe(`${labelText}-error`);
});

[
  {
    authenticate: () => {
      throw new InvalidEmailError();
    },
    errorId: "email-error",
    errorMessage: "emailNotInOurSystem",
  },
  {
    authenticate: () => {
      throw new InvalidPasswordError();
    },
    errorId: "password-error",
    errorMessage: "passwordNotInOurSystem",
  },
  {
    authenticate: () => {
      throw new Error();
    },
    errorId: "log-state",
    errorMessage: "systemError",
  },
].forEach(({ authenticate, errorId, errorMessage }) => {
  test(`Show ${errorMessage} error on specific authenticate exception`, async () => {
    const { getByLabelText, getByText } = render(
      <App authenticate={authenticate} />
    );
    const email = getByLabelText("email");
    const password = getByLabelText("password");
    const loginButton = getByText("login");

    await changeField(email, "testemail@email.com");
    await changeField(password, "testpassword");
    await clickOnButton(loginButton);

    const error = getByText(errorMessage);
    expect(error).toBeInTheDocument();
    expect(error.id).toBe(errorId);
  });
});
