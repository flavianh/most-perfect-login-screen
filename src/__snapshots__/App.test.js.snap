// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<DocumentFragment>
  <form
    action="#"
    aria-label="loginForm"
    class="pure-form pure-form-stacked"
  >
    <div
      class="pure-control-group"
    >
      <label
        for="email"
      >
        email
      </label>
      <input
        id="email"
        name="email"
        required=""
        type="email"
        value=""
      />
    </div>
    <div
      class="pure-control-group"
    >
      <label
        for="password"
      >
        password
      </label>
      <input
        id="password"
        name="password"
        required=""
        type="password"
        value=""
      />
    </div>
    <div
      class="pure-controls"
    >
      <button
        class="pure-button pure-button-primary"
        type="submit"
      >
        login
      </button>
    </div>
    <div>
      loggedOut
    </div>
  </form>
</DocumentFragment>
`;
